### PR TITLE
ps2 menu, memorycard, pad and god knows what else

### PR DESCRIPTION
**Pull request by Fire-Head**

### CODE REVIEW

File: src/core/Pad.cpp [hunk 1]
```diff
@@ -353,24 +369,24 @@ CControllerState::Clear(void)
 
 void CKeyboardState::Clear()
 {
-	for ( int32 i = 0; i < 12; i++  )
+	for ( int32 i = 0; i < 12; i++	)
 		F[i] = 0;
```
Comment (ShFil119) [inline src/core/Pad.cpp:373]: btw can be array size macro

File: src/core/Pad.cpp [hunk 2]
```diff
@@ -353,24 +369,24 @@ CControllerState::Clear(void)
...
-	
-	for ( int32 i = 0; i < 256; i++  )
+
+	for ( int32 i = 0; i < 256; i++	 )
 		VK_KEYS[i] = 0;
```
Comment (ShFil119) [inline src/core/Pad.cpp:376]: also

File: src/core/Pad.cpp [hunk 3]
```diff
@@ -1089,12 +1111,269 @@ void CPad::ProcessPCSpecificStuff(void)
...
+
+#ifdef GTA_PS2
+	bObsoleteControllerMessage = false;
+
+	//int iPressureBtn;
+	int id;
+	int ext_id=0;
+	int state;
+	int rterm_id = 0;
+	unsigned short paddata, tpad;
```
Comment (ShFil119) [inline src/core/Pad.cpp:1126]: imho it's best to avoid `short` it's platform dependent

File: src/save/MemoryCard.cpp [hunk 4]
```diff
@@ -0,0 +1,3082 @@
...
+	return RES_SUCCESS;
+#endif
+}
+
+int32
+CMemoryCard::LoadIconFiles(int32 cardID, char *icon_one, char *icon_two, char *icon_three)
+{
+#if defined(PS2)
+	const uint32 size = 50968;
+	uint8 *data = new uint8[size];
```
Comment (ShFil119) [inline src/save/MemoryCard.cpp:995]: do we need dynamic allocation?

File: src/text/Text.cpp [hunk 5]
```diff
@@ -298,12 +298,32 @@ UnicodeToAsciiForSaveLoad(wchar *src)
 	return aStr;
 }
 
+char*
+UnicodeToAsciiForMemoryCard(wchar *src)
+{
+	static char aStr[256];
+	int len;
+	for(len = 0; *src != '\0' && len < 256-1; len++, src++)
```
Comment (ShFil119) [inline src/text/Text.cpp:306]: array size

File: src/core/FrontendTriggers.h [hunk 6]
```diff
@@ -0,0 +1,1393 @@
...
+		}
+	}
+}
+
+void
+TriggerSaveZone_SaveGameSelect(CMenuMultiChoiceTriggered *widget)
+{	
+	if ( widget )
+	{
+		FillMenuWithMemCardFileListing(&MenuSaveZoneSSL_1, TriggerSaveZone_BackToMainMenuTwoLines, TriggerSaveZone_SaveSlots, NULL, 0, 34, 22);
```
Comment (ShFil119) [inline src/core/FrontendTriggers.h:618]: `NULL`-> `nil`
it's the same, but nicer for spotting usages and for refactoring

File: src/core/FrontendTriggers.h [hunk 7]
```diff
@@ -0,0 +1,1393 @@
...
+		
+		memset(line, 0, sizeof(line));
+		memset(uline, 0, sizeof(uline));
+
+		if (stat2)
+		{
+			if ( itsFloat )
+				sprintf(line, "  %.2f %s %.2f", *(float*)stat, UnicodeToAscii(TheText.Get("FEST_OO")), *(float*)stat2);
+			else
+				sprintf(line, "  %d %s %d", *(int*)stat, UnicodeToAscii(TheText.Get("FEST_OO")), *(int*)stat2);
```
Comment (ShFil119) [inline src/core/FrontendTriggers.h:90]: `int32`?

File: src/core/FrontendTriggers.h [hunk 8]
```diff
@@ -0,0 +1,1393 @@
...
+		}
+		else
+		{
+			if (itsFloat)
+				sprintf(line, "  %.2f", *(float*)stat);
+			else
+				sprintf(line, "  %d", *(int*)stat);
```
Comment (ShFil119) [inline src/core/FrontendTriggers.h:97]: also

File: src/save/MemoryCard.cpp (hunk 4 above)
Comment (Fire-Head) [reply src/save/MemoryCard.cpp:995]: it's by R*, who are you to judge them

File: src/core/Pad.cpp (hunk 3 above)
Comment (Fire-Head) [reply src/core/Pad.cpp:1126]: this one is ps2 platform specific, so doesn't matter

File: src/core/Frontend_PS2.cpp [hunk 9]
```diff
@@ -0,0 +1,3034 @@
...
+CRGBA TITLE_TEXT_COLOR(170, 130, 50, 255);
+CRGBA TEXT_SHADOW_COLOR(0, 0, 0, 255);
+CVector2D SHADOW_VECTOR(1.0f, 1.0f);
+CRGBA SLIDER_RIGHT_COLOR(20, 94, 136, 255);
+CRGBA SLIDER_LEFT_COLOR(86, 196, 255, 255);
+CRGBA MENU_SELECTED_COLOR(255, 212, 88, 255);
+CRGBA rgbaATC(96, 96, 96, 255);	// active text color. not constant
+
+float BUTTONTAB_TEXT_X_SCALES[NUM_PAGES] = { 1.0f };
+float PANEL_TEXT_X_SCALES[NUM_PAGES] = { 1.0f };
```
Comment (ShFil119) [inline src/core/Frontend_PS2.cpp:82]: iirc only first is `1.0f`, it's wanted behavior?

Comment (Fire-Head) [reply src/core/Frontend_PS2.cpp:82]: 003FA420 BUTTONTAB_TEXT_X_SCALES:.float 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0
003FA440 PANEL_TEXT_X_SCALES:.word 0x3F800000, 0, 0, 0, 0, 0, 0

File: src/text/Text.cpp [hunk 10]
```diff
@@ -271,9 +271,9 @@ UnicodeToAscii(wchar *src)
 {
 	static char aStr[256];
 	int len;
-	for(len = 0; *src != '\0' && len < 256-1; len++, src++)
+	for(len = 0; *src != '\0' && len < ARRAY_SIZE(aStr)-1; len++, src++)
```
Comment (Sergeanur) [inline src/text/Text.cpp:274]: Should stay 256

File: src/text/Text.cpp [hunk 11]
```diff
@@ -289,8 +289,22 @@ UnicodeToAsciiForSaveLoad(wchar *src)
 {
 	static char aStr[256];
 	int len;
-	for(len = 0; *src != '\0' && len < 256-1; len++, src++)
-		if(*src < 256)
+	for(len = 0; *src != '\0' && len < ARRAY_SIZE(aStr)-1; len++, src++)
+		if(*src < ARRAY_SIZE(aStr))
```
Comment (Sergeanur) [inline src/text/Text.cpp:293]: Same

Comment (Sergeanur) [reply src/text/Text.cpp:293]: except for line 292, idk why it highlighted xD

File: src/text/Text.cpp (hunk 10 above)
Comment (Sergeanur) [reply src/text/Text.cpp:274]: Wrong string, guuuuuuh

File: src/text/Text.cpp [hunk 12]
```diff
@@ -271,9 +271,9 @@ UnicodeToAscii(wchar *src)
 {
 	static char aStr[256];
 	int len;
-	for(len = 0; *src != '\0' && len < 256-1; len++, src++)
+	for(len = 0; *src != '\0' && len < ARRAY_SIZE(aStr)-1; len++, src++)
 #ifdef MORE_LANGUAGES
-		if(*src < 128 || ((CGame::russianGame || CGame::japaneseGame) && *src < 256))
+		if(*src < 128 || ((CGame::russianGame || CGame::japaneseGame) && *src < ARRAY_SIZE(aStr)))
```
Comment (Sergeanur) [inline src/text/Text.cpp:276]: Should stay 256

File: src/text/Text.cpp (hunk 11 above)
Comment (Sergeanur) [inline src/text/Text.cpp:293]: Same

File: src/core/main.cpp [hunk 13]
```diff
@@ -1773,8 +1803,6 @@ void GameInit()
 	}
 }
 
-// Not used anyway. PS2 main() port
-#ifdef _WIN32
 int
 main(int argc, char *argv[])
```
Comment (erorcun) [inline src/core/main.cpp:1807]: This is needed on Linux, it can't read WinMain so WinMain is renamed to main() in there

Comment (Fire-Head) [reply src/core/main.cpp:1807]: Nope, you are wrong.

https://github.com/GTAmodding/re3/blob/1c11a8081f246dda91ff11d4207d87bb93260b1a/src/core/main.cpp#L1294
https://github.com/GTAmodding/re3/blob/1c11a8081f246dda91ff11d4207d87bb93260b1a/src/core/main.cpp#L1878

**Pull request merged**